bftengine/src/bftengine/MessageBase.cpp

```diff
@@ -12,183 +12,226 @@
+  if (actualSize) *actualSize = 0;
 
-		MessageBase* MessageBase::createObjAndMsgFromLocalBuffer(char* buffer, size_t bufferLength, size_t* actualSize)
-		{
-			if(actualSize) *actualSize = 0;
+  if (bufferLength <= sizeof(RawHeaderOfObjAndMsg)) return nullptr;
 
-			if (bufferLength <= sizeof(RawHeaderOfObjAndMsg)) return nullptr;
+  RawHeaderOfObjAndMsg *pHeader = (RawHeaderOfObjAndMsg *) buffer;
+  if (pHeader->magicNum != magicNumOfRawFormat) return nullptr;
+  if (pHeader->msgSize == 0) return nullptr;
+  if (pHeader->msgSize > maxExternalMessageSize) return nullptr;
+  if (pHeader->msgSize + sizeof(RawHeaderOfObjAndMsg) > bufferLength)
+    return nullptr;
 
-			RawHeaderOfObjAndMsg* pHeader = (RawHeaderOfObjAndMsg*)buffer;
-			if (pHeader->magicNum != magicNumOfRawFormat) return nullptr;
-			if (pHeader->msgSize == 0) return nullptr;
-			if (pHeader->msgSize > maxExternalMessageSize) return nullptr;
-			if (pHeader->msgSize + sizeof(RawHeaderOfObjAndMsg) > bufferLength) return nullptr;
-			
-			char* pBodyInBuffer = buffer + sizeof(RawHeaderOfObjAndMsg);
+  char *pBodyInBuffer = buffer + sizeof(RawHeaderOfObjAndMsg);
 
-			void* msgBody = std::malloc(pHeader->msgSize);
-			memcpy(msgBody, pBodyInBuffer, pHeader->msgSize);
+  void *msgBody = std::malloc(pHeader->msgSize);
+  memcpy(msgBody, pBodyInBuffer, pHeader->msgSize);
 
-			MessageBase* msgObj =
-				new MessageBase(pHeader->sender, (MessageBase::Header*)msgBody, pHeader->msgSize, true);
+  MessageBase *msgObj =
+      new MessageBase(pHeader->sender,
+                      (MessageBase::Header *) msgBody,
+                      pHeader->msgSize,
+                      true);
 
-			if (actualSize) *actualSize = (pHeader->msgSize + sizeof(RawHeaderOfObjAndMsg));
+  if (actualSize)
+    *actualSize = (pHeader->msgSize + sizeof(RawHeaderOfObjAndMsg));
 
-			return msgObj;
-		}
+  return msgObj;
+}
 
+bool MessageBase::equals(const MessageBase &other) const {
+  bool equals = (other.msgSize_ == msgSize_ && other.storageSize_ == storageSize_ &&
+      other.sender_ == sender_ && other.owner_ == owner_);
+  if (!equals)
+    return false;
+  return (memcmp(other.msgBody_, msgBody_, msgSize_) == 0);
+}
 
-	}
+size_t MessageBase::serializeMsg(char *&buf, MessageBase *msg) {
+  // As messages could be empty (nullptr), an additional flag is required to
+  // distinguish between empty and full ones.
+  bool msgEmptyFlag = (msg == nullptr);
+  uint32_t msgEmptyFLagSize = sizeof(msgEmptyFlag);
+  std::memcpy(buf, &msgEmptyFlag, msgEmptyFLagSize);
+  buf += msgEmptyFLagSize;
+
+  size_t actualMsgSize = 0;
+  if (msg) {
+    uint32_t msgSize = msg->sizeNeededForObjAndMsgInLocalBuffer();
+    msg->writeObjAndMsgToLocalBuffer(buf, msgSize, &actualMsgSize);
+    Assert(actualMsgSize != 0);
+    buf += actualMsgSize;
+  }
+  return msgEmptyFLagSize + actualMsgSize;
+}
+
+MessageBase *MessageBase::deserializeMsg(char *&buf, size_t bufLen, size_t &actualSize) {
```

Comment:
Why are you asking? There is no real error handling for now in all bft-engine code. This should be taken as a separate task.

Reply:
bft-engine error handling is implemented via Asserts because in most of the cases there is nothing to do with errors.

Reply:
> There is no real error handling for now in all bft-engine code. This should be taken as a separate task.

If I might interject a personal opinion, I think it might be best if we wrote all new code with error handling to the greatest extent possible rather than marginally complicating the already-daunting task of adding error handling in general.